gather_istio.sh

```diff
@@ -0,0 +1,244 @@
+    else
+      oc adm inspect "--dest-dir=${BASE_COLLECTION_PATH}" "${resource}"
+    fi
+  fi
+}
+
+function inspectNamespace() {
+  local ns
+  ns=$1
+
+  inspect "ns/$ns"
+  for crd in $crds; do
+    inspect "$crd" "$ns"
+  done
+  inspect net-attach-def,roles,rolebindings "$ns"
+}
+
+function main() {
+  local crds
+  echo
+  echo "Executing Istio gather script"
+  echo
+
+  versionFile="${BASE_COLLECTION_PATH}/version"
+  echo "openshift-service-mesh/must-gather"> "$versionFile"
+  version >> "$versionFile"
+
+  # set global variable which is used when calling 'oc adm inspect'
+  get_log_collection_args
+
+  # TODO: add new name label to servicemesh-operator3 pod and use that instead of the control-plane label
+  operatorNamespace=$(oc get pods --all-namespaces -l control-plane=servicemesh-operator3 -o jsonpath="{.items[0].metadata.namespace}")
+  # this gets also logs for all pods in that namespace
+  inspect "ns/$operatorNamespace"
+  inspect clusterserviceversion "${operatorNamespace}"
+
+  inspect nodes
+
+  # TODO: we should probably add a new label or start using 'install.operator.istio.io/owning-resource'
+  # another option is to use 'metadata.ownerReferences' but having a label would be easier for a query
+  for r in $(oc get clusterroles,clusterrolebindings -l install.operator.istio.io/owning-resource -oname); do
+    inspect "$r"
+  done
+  for r in $(oc get clusterroles,clusterrolebindings -l 'app in (istiod,istio-reader)' -oname); do
+    inspect "$r"
+  done
+
+  # inspect all istio.io CRDs
+  crds="$(getCRDs)"
+  for crd in ${crds}; do
+    inspect "crd/${crd}"
+  done
+
+  # inspect all controlled mutatingwebhookconfiguration
+  for mwc in $(oc get mutatingwebhookconfiguration -l app=sidecar-injector -o name);do
+    inspect "${mwc}"
+  done
+  # inspect all controlled validatingwebhookconfiguration
+  for vwc in $(oc get validatingwebhookconfiguration -l app=istiod -o name);do
+    inspect "${vwc}"
+  done
+
+  # inspect Istio and IstioRevision resources, this will just store all CRs as those are cluster-scoped resources
+  inspect "Istio"
+  inspect "IstioRevision"
+
+  # inspect all namespaces with Istio components
+  controlPlanes=$(oc get IstioRevision -o custom-columns=NAME:spec.namespace --no-headers | sort -u)
+  for cp in ${controlPlanes}; do
+    echo
+    echo "Processing control plane namespace: ${cp}"
+
+    inspectNamespace "$cp"
+  done
+
+  # iterate over all Istio revisions
+  for ir in $(oc get IstioRevision -o jsonpath="{.items[*].metadata.name}"); do
+    echo
+    echo "Inspecting ${ir} IstioRevision"
+    cpNamespace=$(oc get IstioRevision "${ir}" -o jsonpath="{.spec.namespace}")
+
+    getSynchronization "${cpNamespace}" "${ir}"
+
+    # iterate over all namespaces which have pods with proxy pointing to this control plane revision
+    # TODO: this might be very resource demanding and causing performance problems in clusters with a lot of pods
```

Comment:
@luksa What do you think about possible perf problem here?

Reply:
Not sure. We had the same thing in 2.x, right? We added all namespaces that were members. We're doing pretty much the same thing here. 

But for the most part, when debugging customer issues, we only need the one namespace that contains the workload that was having the issue. Can we have a cmd-line argument that allows users to specify whether to collect all workload namespaces or just specific ones (this would be the default)?

Reply:
Yes, the goal is the same like in 2.x. But steps to get there are different. In 2.x, we can use `ServiceMeshMemberRoll` to find all namespaces and then we can simply iterate over them.
In 3.x, there is no easy way to get all namespaces which have pods with proxy pointing to given control plane revision. So the command which I currently have (`oc get pods -A -o=jsonpath="{.items[?(@.metadata.annotations.istio\.io/rev==\"${ir}\")].metadata.namespace}`) should do following on the background:

1. get all pods from all namespaces (oc get pods -A)
2. get namespaces for pods which have `metadata.annotations.istio.io/rev==<myRevision>`

First step should be safe as by default `--chunk-size=500` is used. But second part (filtering) will happen in memory on the client side (kubectl/oc) which could cause problems when working with huge clusters.

Regarding 'cmd-line argument that allows users to specify whether to collect all workload namespaces or just specific ones' - looking at `oc adm must-gather -h` it does not allow to specify some extra user defined arguments. It should be possible to pass `myspecial-command.sh` script where we could instruct users to define some env variables but it's ugly solution I think. And it still does not prevent users to purposely say to collect everything.

Reply:
Oh, I don't think kubectl/oc would have trouble with this.

Reply:
Thanks